test/integration_test/action_test.go

```diff
@@ -0,0 +1,223 @@
+
+		defer func() {
+			err := setSourceKubeConfig()
+			require.NoError(t, err, "failed to set kubeconfig to source cluster: %v", err)
+		}()
+	} else {
+		logrus.Infof("creating secret in source")
+	}
+
+	_, err := core.Instance().CreateSecret(secret)
+	if !errors.IsAlreadyExists(err) {
+		require.NoError(t, err, "failed to create secret for volumes")
+	}
+}
+
+func deleteNamespace(t *testing.T, namespace string, cluster string) {
+	if cluster == "dest" {
+		logrus.Infof("deleting namespace %v in destination", namespace)
+		err := setDestinationKubeConfig()
+		require.NoError(t, err, "failed to set kubeconfig to destination cluster: %v", err)
+
+		defer func() {
+			err := setSourceKubeConfig()
+			require.NoError(t, err, "failed to set kubeconfig to source cluster: %v", err)
+		}()
+	} else {
+		logrus.Infof("deleting namespace %v in source", namespace)
+	}
+
+	err := core.Instance().DeleteNamespace(namespace)
+	if err != nil {
+		logrus.Infof("Error deleting namespace: %v\n", err)
+	}
+	time.Sleep(10 * time.Second)
+}
+
+func actionFailoverTest(t *testing.T) {
+
+	appKey := "mysql-enc-pvc"
+	instanceID := "mysql-action"
+	migrationAppId := "mysql-action-migration"
+	actionName := "mysql-action-failover"
+
+	namespace := fmt.Sprintf("%v-%v", appKey, instanceID)
+	deleteNamespace(t, namespace, "src")
+	deleteNamespace(t, namespace, "dest")
+
+	// starts the app on src,
+	// sets cluster pair,
+	// creates a migration
+	ctxs, preMigrationCtx := triggerMigration(
+		t,
+		instanceID,
+		appKey,
+		nil,
+		[]string{migrationAppId},
+		true,
+		true,
+		false,
+		false,
+		"",
+		nil,
+	)
+
+	// validate the following
+	// - migration is successful
+	// - app doesn't start on dest
+	validateAndDestroyMigration(
+		t,
+		ctxs,
+		preMigrationCtx,
+		true,
+		false,
+		true,
+		true,
+		true,
+	)
+
+	// extract migrationObj from specList
+	var migrationObj *v1alpha1.Migration
+	var ok bool
+	for _, specObj := range ctxs[0].App.SpecList {
+		if migrationObj, ok = specObj.(*v1alpha1.Migration); ok {
+			break
+		}
+	}
+
+	err := setSourceKubeConfig()
+	require.NoError(t, err, "failed to set kubeconfig to source cluster: %v", err)
+
+	expectedResources := uint64(4) // 1 sts, 1 service, 1 pvc, 1 pv
+	expectedVolumes := uint64(0)   // 0 volume
+	// validate the migration summary based on the application specs that were deployed by the test
+	validateMigrationSummary(t, preMigrationCtx, expectedResources, expectedVolumes, migrationObj.Name, migrationObj.Namespace)
+
+	scaleFactor := scaleDownApps(t, ctxs)
+	logrus.Infof("scaleFactor: %v", scaleFactor)
+
+	// TODO(dgoel): add a wrapper method that sets and unsets destKubeConfig
+	// trigger action
+	err = setDestinationKubeConfig()
+	require.NoError(t, err, "failed to set kubeconfig to destination cluster: %v", err)
+
+	createActionCR(t, actionName, namespace, ctxs[0])
+
+	err = schedulerDriver.WaitForRunning(preMigrationCtx, defaultWaitTimeout, defaultWaitInterval)
```

Comment:
I think `preMigrationCtx` wont have the the `ActionCR` we should pass `ctxs[0]` here instead if we need to validate ActionCR as well here.